nexus/src/app/instance_platform.rs

```diff
@@ -0,0 +1,481 @@
+            let device = ComponentV0::VirtioNic(VirtioNic {
+                backend_id: SpecKey::Uuid(nic.id),
+                interface_id: nic.id,
+                pci_path: get_pci_bdf(nic.slot, PciDeviceKind::Nic)?,
+            });
+
+            // Sled-agent creates OPTE ports during instance startup using its
+            // per-sled port allocator, so it needs to (and will) fill in the
+            // correct port name once one is assigned.
+            let backend =
+                ComponentV0::VirtioNetworkBackend(VirtioNetworkBackend {
+                    vnic_name: "".to_string(),
+                });
+
+            // N.B. It's crucial to use the NIC ID as the backend name here so
+            // that sled-agent can correlate this component entry with the OPTE
+            // port it created for the guest network interface with this ID.
+            self.add(nic.id.to_string(), backend)?;
+            self.add(device_name, device)?;
+        }
+
+        Ok(())
+    }
+
+    /// Adds a cloud-init disk to this component manifest. The disk will expose
+    /// the supplied `ssh_keys` to the guest.
+    fn add_cloud_init(
+        &mut self,
+        instance: &db::model::Instance,
+        ssh_keys: &[db::model::SshKey],
+    ) -> Result<(), Error> {
+        let keys: Vec<String> =
+            ssh_keys.iter().map(|k| k.public_key.clone()).collect();
+        let base64 = base64::Engine::encode(
+            &base64::engine::general_purpose::STANDARD,
+            instance.generate_cidata(&keys)?,
+        );
+
+        let device = ComponentV0::VirtioDisk(VirtioDisk {
+            backend_id: SpecKey::Name(
+                component_names::CLOUD_INIT_BACKEND.to_string(),
+            ),
+            pci_path: get_pci_bdf(0, PciDeviceKind::CloudInitDisk)
+                .expect("slot 0 is always valid for cloud-init disks"),
+        });
+
+        let backend = ComponentV0::BlobStorageBackend(BlobStorageBackend {
+            base64,
+            readonly: true,
+        });
+
+        self.add(component_names::CLOUD_INIT_DEVICE.to_string(), device)?;
+        self.add(component_names::CLOUD_INIT_BACKEND.to_string(), backend)?;
+
+        Ok(())
+    }
+}
+
+impl super::Nexus {
+    /// Generates a Propolis VM specificaation from the supplied database
+    /// records.
+    pub(crate) async fn generate_vmm_spec(
+        &self,
+        reason: &InstanceRegisterReason,
+        instance: &db::model::Instance,
+        disks: &[db::model::Disk],
+        nics: &[NetworkInterface],
+        ssh_keys: &[db::model::SshKey],
+    ) -> Result<VmmSpec, Error> {
+        let cpus = u8::try_from(instance.ncpus.0.0).map_err(|c| {
+            Error::invalid_value(
+                c.to_string(),
+                "failed to convert instance CPU count to a u8",
+            )
+        })?;
+
+        let mut components = Components::default();
+
+        let mut volumes = vec![];
```

Comment:
also nitpicky: we could save a few potential reallocations by doing
```suggestion
        let mut volumes = Vec::with_capacity(disks.len());
```